### ISSUE TITLE
AssertThat Addition Nullable Value

### ISSUE THREAD

**Issue by gdycus** [jwaliszko/ExpressiveAnnotations#172]
The Children property can be null. If it is null this doesn't evaluate properly. I'm not sure but I assume if Adults is 1 and Children is null then Adults + Children = null. Is there a way to handle this scenario?

[AssertThat("(Adults + Children) <= MaxPeople", ErrorMessage = "Cannot exceed {MaxPeople} people")]

**Comment by gdycus** [comment]
Does anyone have any suggestions? Thanks!

**Comment by jwaliszko** [comment]
It is one of the [null-related traps](https://github.com/jwaliszko/ExpressiveAnnotations#traps).

Take a look at the following expression: `1 + null`.

When evaluated by EA parser, results in `null` (EA server-side parser follows C# rules in that matter). 

![image](https://user-images.githubusercontent.com/1596462/30510774-b4fee6da-9aca-11e7-8bb7-58cab201bf88.png)

Such an expression is then passed to browser. At client-side, there is no separate parser built by EA. Instead, `eval()` method is used. The problem is JavaScript follows different rules when computing such an expression. Evaluation results in `1`.

![image](https://user-images.githubusercontent.com/1596462/30510776-cb058ad8-9aca-11e7-96b4-a081cc00cca3.png)

As a workaround I propose to be explicit of how you treats `null`s. The alternative expression below works coherently at both sides:
```
((Adults == null ? 0 : Adults) + (Children == null ? 0 : Children)) <= MaxPeople
```

**Comment by gdycus** [comment]
I'm getting a parser error

System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.ComponentModel.DataAnnotations.ValidationException: AssertThatValidator: validation applied to Adults field failed. ---> ExpressiveAnnotations.Analysis.ParseErrorException: Parse error on line 1, column 29: ... ? 0 : Children)) <= MaxPeople ... ^--- Argument types must match.

[AssertThat("(Adults + (Children == null ? 0 : Children)) <= MaxPeople", ErrorMessage = "Cannot exceed {MaxPeople} people")]
public short Adults { get; set; }
[AssertThat("(Adults + (Children == null ? 0 : Children)) <= MaxPeople", ErrorMessage = "Cannot exceed {MaxPeople} people")]
public short? Children { get; set; }
public int MaxPeople { get; set; }

**Comment by jwaliszko** [comment]
Indeed, implicit types conversion has been missing for operands of conditional operator. The implementation is provided with the changeset https://github.com/jwaliszko/ExpressiveAnnotations/commit/fa765b320de5aa2638d75f6198a4b3d99e220b95. Will be shipped within next version.

**Comment by gdycus** [comment]
How long before the next release? Thanks

On Sep 16, 2017, at 6:08 PM, Jarosław Waliszko <notifications@github.com<mailto:notifications@github.com>> wrote:


Indeed, implicit types conversion has been missing for operands of conditional operator. The implementation is provided with the changeset fa765b3<https://github.com/jwaliszko/ExpressiveAnnotations/commit/fa765b320de5aa2638d75f6198a4b3d99e220b95>. Will be shipped within next version.

**Comment by jwaliszko** [comment]
Most likely it should be released somewhere this week.

**Comment by jwaliszko** [comment]
Released with [v2.9.6](https://github.com/jwaliszko/ExpressiveAnnotations/releases/tag/v2.9.6)
